src/repo/fs.rs

```diff
@@ -5,147 +5,462 @@ use async_trait::async_trait;
 use bitswap::Block;
 use cid::Cid;
 use core::convert::TryFrom;
-use futures::lock::Mutex;
-use futures::stream::StreamExt;
-use std::collections::HashSet;
-use std::ffi::OsStr;
+use std::collections::HashMap;
+use std::io::Read;
 use std::path::PathBuf;
-use tokio::{
-    fs,
-    io::{AsyncReadExt, AsyncWriteExt},
-};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::sync::Mutex;
+use tracing_futures::Instrument;
+
+use tokio::fs;
+use tokio::sync::broadcast;
 
 use super::{BlockRm, BlockRmError, RepoCid};
+use std::hash::Hash;
+
+// mod flatfs;
+type ArcMutexMap<A, B> = Arc<Mutex<HashMap<A, B>>>;
 
 #[derive(Debug)]
 pub struct FsBlockStore {
+    /// The base directory under which we have a sharded directory structure, and the individual
+    /// blocks are stored under the shard. See unixfs/examples/cat.rs for read example.
     path: PathBuf,
-    cids: Mutex<HashSet<RepoCid>>,
+
+    /// Synchronize concurrent reads and writes to the same Cid.
+    /// If the write ever happens, the message sent will be Ok(()), on failure it'll be an Err(()).
+    /// Since this is a broadcast channel, the late arriving receiver might not get any messages.
+    writes: ArcMutexMap<RepoCid, broadcast::Sender<Result<(), ()>>>,
+
+    /// Initially used to demonstrate a bug, not really needed anymore. Could be used as a basis
+    /// for periodic synching to disk to know much space we have used.
+    written_bytes: AtomicU64,
```

Comment:
This is interesting in terms of stats - maybe eventually move it to some `RepoStat` struct but for now probably fine to leave